libp2p/src/builder.rs

```diff
@@ -0,0 +1,1795 @@
+            phantom: PhantomData,
+            phase: RelayPhase {
+                transport: libp2p_dns::TokioDnsConfig::system(self.phase.transport)?,
+            },
+        })
+    }
+}
+
+impl<Provider, T> SwarmBuilder<Provider, DnsPhase<T>> {
+    fn without_dns(self) -> SwarmBuilder<Provider, RelayPhase<T>> {
+        SwarmBuilder {
+            keypair: self.keypair,
+            phantom: PhantomData,
+            phase: RelayPhase {
+                transport: self.phase.transport,
+            },
+        }
+    }
+}
+
+// Shortcuts
+impl<Provider, T: AuthenticatedMultiplexedTransport> SwarmBuilder<Provider, DnsPhase<T>> {
+    pub fn with_behaviour<B, R: TryIntoBehaviour<B>>(
+        self,
+        constructor: impl FnOnce(&libp2p_identity::Keypair) -> R,
+    ) -> Result<SwarmBuilder<Provider, SwarmPhase<T, B>>, R::Error> {
+        self.without_dns()
+            .without_relay()
+            .without_websocket()
+            .with_behaviour(constructor)
+    }
+}
+
+pub struct RelayPhase<T> {
+    transport: T,
+}
+
+// TODO: Noise feature or tls feature
+#[cfg(feature = "relay")]
+impl<Provider, T> SwarmBuilder<Provider, RelayPhase<T>> {
+    // TODO: This should be with_relay_client.
+    pub fn with_relay(self) -> SwarmBuilder<Provider, RelayTlsPhase<T>> {
+        SwarmBuilder {
+            keypair: self.keypair,
+            phantom: PhantomData,
+            phase: RelayTlsPhase {
+                transport: self.phase.transport,
+            },
+        }
+    }
+}
+
+pub struct NoRelayBehaviour;
+
+impl<Provider, T> SwarmBuilder<Provider, RelayPhase<T>> {
+    fn without_relay(self) -> SwarmBuilder<Provider, WebsocketPhase<T, NoRelayBehaviour>> {
+        SwarmBuilder {
+            keypair: self.keypair,
+            phantom: PhantomData,
+            phase: WebsocketPhase {
+                transport: self.phase.transport,
+                relay_behaviour: NoRelayBehaviour,
+            },
+        }
+    }
+}
+
+// Shortcuts
+impl<Provider, T: AuthenticatedMultiplexedTransport> SwarmBuilder<Provider, RelayPhase<T>> {
+    #[cfg(all(not(target_arch = "wasm32"), feature = "websocket"))]
+    pub fn with_websocket(
+        self,
+    ) -> SwarmBuilder<
+        Provider,
+        WebsocketTlsPhase<impl AuthenticatedMultiplexedTransport, NoRelayBehaviour>,
+    > {
+        self.without_relay().with_websocket()
+    }
+
+    pub fn with_behaviour<B, R: TryIntoBehaviour<B>>(
+        self,
+        constructor: impl FnOnce(&libp2p_identity::Keypair) -> R,
+    ) -> Result<SwarmBuilder<Provider, SwarmPhase<T, B>>, R::Error> {
+        self.without_relay()
+            .without_websocket()
+            .with_behaviour(constructor)
+    }
+}
+
+#[cfg(feature = "relay")]
+pub struct RelayTlsPhase<T> {
+    transport: T,
+}
+
+#[cfg(feature = "relay")]
+impl<Provider, T> SwarmBuilder<Provider, RelayTlsPhase<T>> {
+    #[cfg(all(not(target_arch = "wasm32"), feature = "tls"))]
+    pub fn with_tls(
```

Comment:
I find it a bit confusing that this function is also called `with_tls` but it describes the encryption that is used OVER a relayed connection. I thought I had commented this somewhere but can't find it now. Should we rename this to `with_relayed_tls`?

Reply:
Solved with the recent refactor to use `with_tcp(security_protocol, muxer)`.